lib/queues.js

```diff
@@ -611,5 +610,22 @@ module.exports = class Queues {
     // Lazy load configuration.
     return this._ironium.configPromise;
   }
+
+  // Returns the maximum amount of jobs a queue should try to reserve.
+  // Every queue gets a share of reservation capacity to better deal
+  // with load spikes across queues.
+  getReserveCapacity() {
+    return this.configPromise
+      .then(config => {
+        const totalCapacity       = config.concurrency - this._activeJobs;
+        const maxCapacityPerQueue = Math.floor(totalCapacity / this._queues.size);
+        const capacity            = Math.min(totalCapacity, maxCapacityPerQueue);
```

Comment:
> then a spike of n jobs on each queue would cause actual concurrency to be concurrency * queue count.

Since reserving is picking whatever jobs are in the queue, there's no algorithm that can do that. The only way to enforce a concurrency limit is to reserve, and if there are excess jobs, un-reserve some of those.

The best an algorithm can do is minimize how many reserved jobs are un-reserved, it could do that by reserving less than full capacity (say max(50%, 1)).